### PR TITLE
A better fix for #tab-titles-are-too-long

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request
  Add a `startingTitle` setting. This will act just like setting the `lpTitle`
  member of a `STARTUPINFO`, which is similar to how cmd and powershell being
  started from a shortcut works today. `cmd` and `powershell` will read their
  initial title and keep it around, instead of setting the title like `bash`
  does. When commands in `cmd` are executed, cmd will append them to the end of
  this value, like cmd currently does when started from the "Command Prompt"
  shortcut.

  Also adds default `startingTitle`s to `cmd`, `powershell`, and `pwsh`.

  This is accomplished with a change to conhost.exe to plumb the `lpTitle` down
  to the child process it's starting. This shouldn't necessarily regress
  anything, since starting conhost directly and having conhost start a
  executable on the caller's behalf is a relatively new and undocumented
  scenario.

Here's what it looks like for cmd, powershell, and ubuntu. I set the Ubuntu `startingTitle` to "Ubuntu", which was immediately overridden when `bash` started. Cmd is running `ping -t github.com`, which you can see appended to the tab title like usual.
image

## References
literally every issue that mentions the length of the tab title, but #2327 was the one that broke the camel's back.

Also is a way better solution than #2348.

## PR Checklist
* [x] Closes #the-endless-stream-of-complaints-about-tab-titles
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [x] Requires documentation to be updated
* [x] I discussed this with Dustin yesterday. Turns out it was _very_ doable.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
We need to consider what this means for CreatePseudoConsole (CreatePseudoConsoleExW)

**Comment by zadjii-msft** [comment]
@DHowett-MSFT This will be easy for CreateaPseudoConsole, since _we'll_ be creating the child process.

File: doc/cascadia/SettingsSchema.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ Properties listed below are specific to each unique profile.
 | `padding` | _Required_ | String | `0, 0, 0, 0` | Sets the padding around the text within the window. Can have three different formats: `"#"` sets the same padding for all sides, `"#, #"` sets the same padding for left-right and top-bottom, and `"#, #, #, #"` sets the padding individually for left, top, right, and bottom. |
 | `snapOnInput` | _Required_ | Boolean | `true` | When set to `true`, the window will scroll to the command input line when typing. When set to `false`, the window will not scroll when you start typing. |
 | `startingDirectory` | _Required_ | String | `%USERPROFILE%` | The directory the shell starts in when it is loaded. |
+| `startingTitle` | Optional | String | | The title to pass to the shell on startup. Some shells (like `bash`) may choose to ignore this initial value, while others (`cmd`, `powershell`) may use this value over the lifetime of the application. |
```
Comment (carlos-zamora) [inline doc/cascadia/SettingsSchema.md]: default should be empty string

Should we add a note at the bottom of the doc saying what the defaults are for the generated profiles? Or maybe that's something for a completely different user doc?

Comment (carlos-zamora) [reply doc/cascadia/SettingsSchema.md]: Also maybe specifying the specific difference between `startingTitle` and `tabTitle`?

Comment (DHowett-MSFT) [reply doc/cascadia/SettingsSchema.md]: Default is _unset_, not _empty string_. Distinction!

**Comment by DHowett-MSFT** [comment]
I think we should rationalize the names `startingTitle` and `tabTitle` before shipping this. I always thought `tabTitle` should be `tabTitleOverride`... because it honestly makes it sound like `tabTitle` _could be replaced with this new behavior_

File: src/cascadia/TerminalConnection/ConhostConnection.cpp [hunk 2]
```diff
@@ -90,6 +92,19 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
 
         THROW_LAST_ERROR_IF_NULL(_hOutputThread);
 
+        STARTUPINFO si = { 0 };
+        si.cb = sizeof(STARTUPINFOW);
+
+        // If we have a startingTitle, create a mutable character buffer to add
+        // it to the STARTUPINFO.
+        std::unique_ptr<wchar_t[]> mutableTitle = std::make_unique<wchar_t[]>(_startingTitle.size() + 1);
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalConnection/ConhostConnection.cpp]: we'll allocate one byte and then discard it later if the user doesn't set a starting title.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT I'm also certainly not sure on the name. I was originally going to go with "shortcutTitle" but that seemed ridiculous when I wrote it - no one outside the console team understands console shortcuts well. 

"initialTitle" and "startingTitle" both seem pretty similar to me, but I'm not a PM. Words aren't my job. Summoning @cinnamon-msft and @bitcrazed. 

Frankly, until #1320 lands, I think that `tabTitle` is a strictly worse option than `startingTitle`, but once #1320 _does_ land, the two might be very powerful in conjunction. I'm going to await some guidance from our PM colleagues on naming here.

**Comment by DHowett-MSFT** [comment]
Controversial: what if we just always push the profile name into Conhost? Maybe we don’t need a new setting for this particular feature. Consider that the profile is the same as the .LNK and previously we used the LNK name as the title?

**Comment by DHowett-MSFT** [comment]
Oh oh and we then make it so that tabTitle _is_ this setting for if you want it to be different from the profile. We would get rid of the Overriding aspect (which I think is a feature that nobody truly wanted when it was only a static name)?

**Comment by zadjii-msft** [comment]
@DHowett-MSFT I don't hate the idea of using the profile.name as the starting title. I'd be curious if anyone wanted a separate value for it, because that would be impossible to do with that implementation.

However, I think that `tabTitle` should maybe stay reserved for #1320, when the overriding will make more sense. Though, we could always add a new setting _then_ if we really wanted.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT @carlos-zamora @cinnamon-msft @bitcrazed @miniksa 
In preparation for today's later discussion on this topic, I tried to write up scenarios and proposed solutions. @ me if I forgot any scenarios here.

### User Stories:

1. A user wants to be able to use the executable path as their starting title
    - Does anyone want this?
2. A user wants to be able to set a custom starting title, but have that title be overridable
3. A user wants to be able to set an overridable starting title, different from the profile name
    - Presumably someone will want this
4. A user totally wants to ignore the VT title and use something else
    - This will make more sense in the post [#1320] "Support runtime variables in the custom user title" settings

### Solutions:

1. `name`, `startingTitle`, `tabTitle` (the PR as it exists currently)
    * a. `name` is only ever used as the profile name.
    * b. If `startingTitle` isn't set, then the executable path is used
    * c. If `startingTitle` is set, it's used as the initial title
    * d. If `tabTitle` is set, it overrides the title from the terminal
    * e. Current users of `tabTitle` need to manually update to the new behavior.
2. `name` as starting title, `tabTitle` as a different starting title
    * a. `name` is used as the starting title and the profile name in the dropdown
    * b. If `tabTitle` is set, we'll use that as the overridable starting title instead.
    * c. In the future, `dynamicTabTitle` or `tabTitleOverride` could be added to support [#1320]
    * d. Current users of `tabTitle` automatically get the new (different!) behavior.
    * e. User Story 1 is impossible
        - Does anyone want the behavior _ever_? Perhaps making that scenario impossible is good?
3. `name` unchanged, `tabTitle` as the starting title
    * a. `name` is only ever used as the profile name.
    * b. If `tabTitle` is set, we'll use that as the overridable starting title.
    * c. In the future, `dynamicTabTitle` or `tabTitleOverride` could be added to support [#1320]
    * d. Current users of `tabTitle` automatically get the new (different!) behavior.

[#1320]: https://github.com/microsoft/terminal/issues/1320

File: src/cascadia/TerminalConnection/ConhostConnection.cpp (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConhostConnection.cpp]: This remains true even in the latest edition.

**Comment by DHowett-MSFT** [comment]
SA build is failing due to disk space.

**Comment by IanKemp** [comment]
Dear god I need this. When will you guys be dropping a new release that contains this functionality?

**Comment by Uxorious** [comment]
> d. We'll add suppressApplicationTitle when someone complains

Not a complaint, but I would love this :-)

**Comment by sth4me** [comment]
Any option can overrides the title Or just always shows in the tab?When I ssh to a server,I want to know server name I custom.

**Comment by comsuoncha** [comment]
@zadjii-msft can you please add suppressApplicationTitle feature? I need it on WSL, my tab titles are too long...

**Comment by zadjii-msft** [comment]
@thongtrinh2204 hey look, it's already in PR  #2814

**Comment by thongtrinh2204** [comment]
> @thongtrinh2204 hey look, it's already in PR #2814

Sounds good. Can wait for the next release.

**Pull request merged**